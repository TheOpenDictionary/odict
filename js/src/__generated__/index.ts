export * from "./service";
export * from "./enums";
