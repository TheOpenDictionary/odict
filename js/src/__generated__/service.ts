// automatically generated by the FlatBuffers compiler, do not modify

export { CompilePayload } from './compile-payload.js';
export { LookupPayload } from './lookup-payload.js';
export { MarkdownStrategy } from './markdown-strategy.js';
export { ODictMethod } from './odict-method.js';
export { SearchPayload } from './search-payload.js';
export { SplitPayload } from './split-payload.js';
export { WritePayload } from './write-payload.js';
