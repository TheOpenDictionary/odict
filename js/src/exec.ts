import { exec as _exec } from "node:child_process";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";

const run = promisify(_exec);

/**
 * Executes the ODict binary
 *
 * @param args Arguments to pass to the executable
 * @returns The raw stdout output string
 */
export async function exec(...args: string[]) {
  let odictExecutable = "odict";

  if (process.env.RUNTIME_ENV === "test") {
    odictExecutable = join(fileURLToPath(import.meta.url), "../../../bin/odict");
  }

  const { stderr, stdout } = await run(
    [odictExecutable, "--quiet", ...args].join(" "),
    { shell: "/bin/bash" }
  );

  if (stderr) {
    throw new Error(stderr);
  }

  return stdout;
}
