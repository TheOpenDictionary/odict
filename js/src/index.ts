export * from "./dictionary.js";
export * from "./types.js";
