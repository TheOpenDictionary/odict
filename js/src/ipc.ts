import { appendFileSync } from "node:fs";
import { v4 as uuid } from "uuid";

import { ChildProcess, spawn } from "node:child_process";
import { EventEmitter } from "node:events";

export interface GoPayload<Data, Error> {
	id: string;
	event: string;
	data: Data;
	error: Error;
	SR: boolean;
}

export type Channel = string | { event: string; id: string };

/**
  Adapted from https://github.com/Akumzy/ipc-node/blob/master/src/index.ts

  MIT License

  Copyright (c) 2019 Akuma Isaac Akuma

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
 */
export class IPC extends EventEmitter {
	go: ChildProcess | null;

	closed: boolean;

	constructor(private binPath: string) {
		super();

		this.go = null;

		this.closed = false;

		/**
		 * The `Golang` process will be pinging at every 20 seconds
		 * and will wait another 20 seconds for reply via `pong` event name
		 * else it will kill it's process.
		 */
		this.on("ping", () => this.send("pong"));
	}

	/**
	 * Start the child process
	 * @param arg
	 */
	public init(arg: string[] = []) {
		const self = this;
		const go = spawn(this.binPath, arg, {});

		this.closed = false;
		this.go = go;

		go.stderr.setEncoding("utf8");
		go.stdout.setEncoding("utf8");

		// emit the errors
		go.stderr.on("error", (e) => self.emit("error", e));
		go.stderr.on("data", (e) => self.emit("log", e));

		let outBuffer = "";

		go.stdout.on("data", (s) => {
			outBuffer += s;

			if (s.endsWith("}\\n")) {
				if (process.env.ODICT_DEBUG_IPC) {
					appendFileSync("ipc.log", `NODE RECEIVED: ${outBuffer}`);
				}
				self._processData(outBuffer);
				outBuffer = "";
			}
		});

		go.once("close", (_) => {
			self.closed = true;
			self.emit("close");
		});

		process.on("beforeExit", () => this.kill());

		return this;
	}

	private _processData(payload: string) {
		const _data = this.parseJSON(payload);

		if (Array.isArray(_data)) {
			for (const item of _data) {
				const { id, error, data, event } = item;

				this.emit("data", item);
				this.emit(event, data, error);
				this.emit(`${event}:${id}`, data, error);
			}
		}
	}

	/**
	 * Kill the child process
	 */
	public kill() {
		try {
			this.send("___EXIT___", null);
			this.closed = true;
			this.go?.kill();
			this.emit("killed");
		} catch (error) {
			console.log(error);
		}
	}

	/**
	 * Send message to `Golang` process
	 * @param event
	 * @param data
	 */
	public send<Data>(event: Channel, data: Data | undefined = undefined) {
		this._send(event, data, false);
	}

	/**
	 * sendRaw gives your access to a third `boolean` argument which
	 * is used to determine if this is a sendAndReceive action
	 */
	public sendRaw<Data>(event: Channel, data: Data, isSendAndReceive = false) {
		this._send(event, data, isSendAndReceive);
	}

	/**
	 *
	 * @param event
	 * @param data
	 * @param SR this tells `Go` process if this message needs an acknowledgement
	 */
	private _send<Data>(event: Channel, data: Data, SR: boolean) {
		try {
			if (this.go && !this.closed && this.go.stdin?.writable) {
				const payload =
					typeof data === "object" || Array.isArray(data)
						? JSON.stringify(data)
						: data;

				// We are converting this to `JSON` this to preserve the
				// data types
				const d = JSON.stringify({
					id: typeof event === "string" ? uuid() : event.id,
					event: typeof event === "string" ? event : event.event,
					data: payload,
					SR: !!SR,
				});

				if (this.go.stdin.writable) {
					const text = d + "\n";

					if (process.env.ODICT_DEBUG_IPC) {
						appendFileSync("ipc.log", `NODE SENT: ${text}`);
					}

					this.go.stdin.write(text);
				}
			}
		} catch (error) {
			this.emit("error", error);
		}
	}

	/**
	 *  Send and receive an acknowledgement through
	 * a callback from `Go` process
	 * @param event
	 * @param data
	 * @param cb
	 */
	public sendAndReceive (
		event: string,
		data: any,
		cb: (error: Error, data: any) => void,
	) {
		const id = uuid();

		this._send({ event, id }, data, true);

		const rc = `${event}:${id}`;

		this.once(rc, (data, error) => {
			if (typeof cb === "function") {
				cb(error, data);
			}
		});
	}

	/**
	 *  Receive and send back acknowledgement/data to `GO`
	 * a callback from `Go` process
	 * @param event
	 * @param data
	 * @param cb
	 */
	public onReceiveAnSend<Data>(
		event: string,
		cb: (channel: Channel, data: Data) => void,
	) {
		const channel = { event, id: uuid() };

		this.on(event, (data) => {
			if (typeof cb === "function") {
				cb(channel, data);
			}
		});
	}

	private parseJSON<Data, Error>(s: string): GoPayload<Data,Error>[] | null {
		try {
			let data = s.replace(/}\\n/g, "},");

			if (data.endsWith(",")) {
				data = data.slice(0, -1).trim();
			}

			return JSON.parse(`[${data}]`);
		} catch (error) {
			this.emit("parse-error", error);
			return null;
		}
	}
}
