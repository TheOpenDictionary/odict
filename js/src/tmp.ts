import { exec as _exec } from "node:child_process";
import { randomBytes } from "node:crypto";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { promisify } from "node:util";

import { exec } from "./exec";
import type { DictionaryOptions } from "./types";

const run = promisify(_exec);

export async function withTemporaryFile<T>(
  cb: (path: string) => Promise<T> | T
): Promise<T> {
  const file = randomBytes(6).toString("hex");
  const tmp = join(tmpdir(), file);
  const value = await cb(tmp);

  await rm(tmp);

  return value;
}
