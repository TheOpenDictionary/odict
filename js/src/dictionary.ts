import { Builder } from "flatbuffers";

import { CompilePayload } from "./__generated__/compile-payload.js";
import {
  LookupPayload,
  MarkdownStrategy,
  SearchPayload,
  SplitPayload,
  WritePayload,
} from "./__generated__/index.js";
import { ODictMethod } from "./__generated__/odict-method.js";
import { startService } from "./service.js";
import type {
  DictionaryOptions,
  Entry,
  LookupOptions,
  Query,
  SearchOptions,
} from "./types.js";
import { generateOutputPath, queryToString } from "./utils.js";

function createBuilder() {
  return new Builder(1024);
}

class Dictionary {
  private readonly options: DictionaryOptions;

  constructor(
    public readonly path: string,
    options: Partial<DictionaryOptions> = {},
  ) {
    this.options = options;
  }

  /**
   * Compiles an ODXML file to a compiled dictionary
   *
   * @param xmlPath The path to the XML file to compile
   * @param outPath The destination path (autogenerated if omitted)
   * @returns A pointer to the compiled Dictionary
   */
  static async compile(xmlPath: string, outPath?: string): Promise<Dictionary> {
    const service = await startService();
    const out = outPath ?? generateOutputPath(xmlPath);
    const builder = createBuilder();

    const pathS = builder.createString(xmlPath);
    const outS = builder.createString(out);

    CompilePayload.startCompilePayload(builder);
    CompilePayload.addPath(builder, pathS);
    CompilePayload.addOut(builder, outS);

    const payload = CompilePayload.endCompilePayload(builder);

    builder.finish(payload);

    await service.run(ODictMethod.Compile, builder.asUint8Array());

    return new Dictionary(out);
  }

  /**
   * Writes an ODXML string to a compiled dictionary file
   *
   * @param xml The XML to compile
   * @param outPath The destination path
   * @returns A pointer to the compiled dictionary
   */
  static async write(xml: string, outPath: string): Promise<Dictionary> {
    const service = await startService();
    const builder = createBuilder();
    const xmlS = builder.createString(xml);
    const outS = builder.createString(outPath);

    WritePayload.startWritePayload(builder);
    WritePayload.addXml(builder, xmlS);
    WritePayload.addOut(builder, outS);

    const payload = WritePayload.endWritePayload(builder);

    builder.finish(payload);

    await service.run(ODictMethod.Write, builder.asUint8Array());

    return new Dictionary(outPath);
  }

  /**
   * Indexes a compiled dictionary so it can be searched via the search() method
   */
  async index() {
    const service = await startService(this.path);

    await service.run(ODictMethod.Index);
  }

  /**
   * Searches a compiled dictionary using full-text search
   *
   * @param query The query (or a list of queries) to search
   * @param options Search options
   * @returns A nested array of entries
   */
  async search(
    query: Query | Query[],
    options: SearchOptions = {},
  ): Promise<Entry[][]> {
    const service = await startService(this.path);
    const queries = Array.isArray(query) ? query : [query];

    return Promise.all(
      queries.map(queryToString).map(async (query) => {
        const builder = createBuilder();
        const queryS = builder.createString(query);

        SearchPayload.startSearchPayload(builder);
        SearchPayload.addQuery(builder, queryS);
        SearchPayload.addExact(builder, options.exact ?? false);
        SearchPayload.addForce(builder, options.force ?? false);

        const payload = SearchPayload.endSearchPayload(builder);

        builder.finish(payload);

        return service.run<Entry[]>(ODictMethod.Search, builder.asUint8Array());
      }),
    );
  }

  /**
   * Returns a list of all headwords in the dictionary
   *
   * @returns A list of all headwords in the dictionary
   */
  async lexicon(): Promise<string[]> {
    const service = await startService(this.path);
    return service.run<string[]>(ODictMethod.Lexicon);
  }

  /**
   * Looks up an entry in the dictionary
   *
   * @param query The query (or a list of queries) to lookup
   * @param options Lookup options
   * @returns A nested array of entries
   */
  async lookup(
    query: Query | Query[],
    options: LookupOptions = {},
  ): Promise<Entry[][]> {
    const service = await startService(this.path);

    const queries = Array.isArray(query) ? query : [query];

    const { follow, split = this.options.defaultSplitThreshold } = options;

    const builder = createBuilder();

    const queriesS = queries
      .map(queryToString)
      .map((str) => builder.createString(str));

    const queriesV = LookupPayload.createQueriesVector(builder, queriesS);

    let markdown: MarkdownStrategy | undefined;

    if (options.markdownStrategy === "html") {
      markdown = MarkdownStrategy.HTML;
    }

    if (options.markdownStrategy === "disable") {
      markdown = MarkdownStrategy.Disable;
    }

    if (options.markdownStrategy === "text") {
      markdown = MarkdownStrategy.Text;
    }

    LookupPayload.startLookupPayload(builder);
    LookupPayload.addQueries(builder, queriesV);
    LookupPayload.addFollow(builder, follow ?? false);
    LookupPayload.addSplit(builder, split ?? 0);

    if (markdown) {
      LookupPayload.addMarkdown(builder, markdown);
    }

    const payload = LookupPayload.endLookupPayload(builder);

    builder.finish(payload);

    return service.run<Entry[][]>(ODictMethod.Lookup, builder.asUint8Array());
  }

  /**
   * Splits a query into its definable subparts
   *
   * @param query The query (or a list of queries) to lookup
   * @param options Lookup options
   * @returns A nested array of entries
   */
  async split(query: string, threshold: number): Promise<Entry[]> {
    const service = await startService(this.path);
    const builder = createBuilder();
    const queryS = builder.createString(query);

    SplitPayload.startSplitPayload(builder);
    SplitPayload.addThreshold(builder, threshold);
    SplitPayload.addQuery(builder, queryS);

    const payload = SplitPayload.endSplitPayload(builder);

    builder.finish(payload);

    return service.run<Entry[]>(ODictMethod.Split, builder.asUint8Array());
  }
}

export { Dictionary };
