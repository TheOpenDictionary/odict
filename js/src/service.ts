import { v4 as uuid } from "uuid";

import { join } from "node:path";
import { fileURLToPath } from "node:url";

import { ODictMethod } from "./__generated__";
import { IPC } from "./ipc";

type ProcessWrapper = {
	run: <T>(method: ODictMethod, buffer?: Uint8Array) => Promise<T>;
	stop: () => void;
};

const openDictionaries: Map<string, ProcessWrapper> = new Map();

export function teardownServices() {
	openDictionaries.forEach((process) => process.stop());
	openDictionaries.clear();
}

process.on("beforeExit", () => {
	teardownServices();
});

function getExecutablePath() {
	return process.env.RUNTIME_ENV === "test"
		? join(fileURLToPath(import.meta.url), "../../../bin/odict")
		: "odict";
}

/**
 * Executes the ODict binary
 *
 * @param args Arguments to pass to the executable
 * @returns The raw stdout output string
 */
export function startService(dictionaryPath?: string) {
	const cacheKey = dictionaryPath ?? "shared";

	if (openDictionaries.has(cacheKey)) {
		return openDictionaries.get(cacheKey) as ProcessWrapper;
	}

	const executable = getExecutablePath();

	const ipc = new IPC(executable);

	ipc.init(["service", dictionaryPath ?? ""]);

	ipc.on("close", () => openDictionaries.delete(cacheKey));

	ipc.on("killed", () => openDictionaries.delete(cacheKey));

	function stop() {
		ipc.kill();
	}

	const processWrapper: ProcessWrapper = {
		stop,
		run(method, payload) {
			return new Promise((resolve, reject) => {
				ipc.sendAndReceive(
					ODictMethod[method],
					payload ? Buffer.from(payload).toString("base64") : "",
					(err, data) => (err ? reject(err) : resolve(data)),
				);
			});
		},
	};

	openDictionaries.set(cacheKey, processWrapper);

	return processWrapper;
}
