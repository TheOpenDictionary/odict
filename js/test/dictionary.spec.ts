import { afterAll, beforeAll, describe, expect, it } from "@jest/globals";
import findProcess from "find-process";

import { existsSync } from "node:fs";
import { rm, stat } from "node:fs/promises";
import { join } from "node:path";
import { fileURLToPath } from "node:url";

import { Dictionary } from "../src";
import { teardownServices } from "../src/service";

describe("Dictionary", () => {
  let dict1: Dictionary;

  afterAll(() => {
    teardownServices();
  });

  beforeAll(async () => {
    dict1 = await Dictionary.compile(
      join(
        fileURLToPath(new URL(import.meta.url)),
        "../../../examples/example1.xml",
      ),
    );

    const stats = await stat(dict1.path);

    expect(stats.isFile).toBeTruthy();
  });

  it("restarts if the process was killed", async () => {
    const result1 = await dict1.lookup({
      word: "dog",
      fallback: "dog",
    });

    expect(result1).toMatchSnapshot();

    const processes = await findProcess("name", "odict", true);

    processes.forEach((p) => process.kill(p.pid));

    await new Promise((r) => setTimeout(r, 0)); // Not sure why this is needed...

    const result2 = await dict1.lookup({
      word: "run",
      fallback: "run",
    });

    expect(result2).toMatchSnapshot();
  });

  it("can lookup terms properly", async () => {
    const result = await dict1.lookup({ word: "cat", fallback: "cat" });
    expect(result).toMatchSnapshot();
  });

  it("doesn't split unless specified", async () => {
    const result = await dict1.lookup("catdog");
    expect(result[0].length).toBe(0);
  });

  it("can return the lexicon", async () => {
    const result = await dict1.lexicon();
    expect(result).toStrictEqual(["cat", "dog", "poo", "ran", "run"]);
  });

  it("can write raw XML", async () => {
    await Dictionary.write(
      '<dictionary><entry term="hello"><ety><sense pos="v"><definition>hello world</definition></sense></ety></entry><entry term="world"><ety><sense pos="v"><definition>hello world</definition></sense></ety></entry></dictionary>"    )',
      "test.odict",
    );

    expect(existsSync("test.odict")).toBeTruthy();

    await rm("test.odict");
  });

  it("can split terms during lookup", async () => {
    const result = await dict1.lookup("catdog", { split: 3 });
    expect(result).toMatchSnapshot();
  });

  it("can split terms properly", async () => {
    const result = await dict1.split("catdog", 2);
    expect(result).toMatchSnapshot();
  });

  it("can index and search a dictionary", async () => {
    await dict1.index();

    const results = await dict1.search("run");

    expect(results).toMatchSnapshot();
  });
});
