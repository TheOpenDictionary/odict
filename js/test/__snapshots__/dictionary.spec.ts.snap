// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dictionary can index and search a dictionary 1`] = `
[
  [
    {
      "etymologies": null,
      "see": "run",
      "term": "ran",
    },
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    "The dog runs after the cat.",
                    "The horse ran away.",
                  ],
                  "value": "(vertebrates) To move swiftly",
                },
                {
                  "value": "Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.",
                },
                {
                  "value": "A pleasure trip.",
                },
                {
                  "value": "Flight, instance or period of fleeing.",
                },
                {
                  "value": "Migration (of fish).",
                },
                {
                  "value": "A group of fish that migrate, or ascend a river for the purpose of spawning.",
                },
              ],
            },
            "v": {
              "groups": [
                {
                  "definitions": [
                    {
                      "examples": [
                        "The dog runs after the cat.",
                        "The horse ran away.",
                      ],
                      "value": "(vertebrates) To move swiftly (figuratively)",
                    },
                    {
                      "value": "(fluids) To flow.",
                    },
                    {
                      "value": "(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled.",
                    },
                    {
                      "value": "(social) To carry out an activity.",
                    },
                    {
                      "value": "To extend or persist, statically or dynamically, through space or time.",
                    },
                    {
                      "value": "(transitive) To execute or carry out a plan, procedure or program.",
                    },
                  ],
                  "description": "A number of verb senses",
                },
              ],
            },
          },
        },
      ],
      "term": "run",
    },
  ],
]
`;

exports[`Dictionary can split terms properly 1`] = `
[
  {
    "etymologies": [
      {
        "description": "Latin root",
        "senses": {
          "n": {
            "definitions": [
              {
                "examples": [
                  "There goes a cat!",
                ],
                "notes": [
                  {
                    "examples": [
                      "Some example",
                    ],
                    "value": "Some definition note",
                  },
                ],
                "value": "a cat",
              },
            ],
          },
        },
      },
    ],
    "term": "cat",
  },
  {
    "etymologies": [
      {
        "description": "Latin root",
        "senses": {
          "un": {
            "definitions": [
              {
                "value": "a dog",
              },
            ],
          },
        },
      },
    ],
    "pronunciation": "dooooog",
    "term": "dog",
  },
]
`;

exports[`Dictionary lookup can split terms 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    "There goes a cat!",
                  ],
                  "notes": [
                    {
                      "examples": [
                        "Some example",
                      ],
                      "value": "Some definition note",
                    },
                  ],
                  "value": "a cat",
                },
              ],
            },
          },
        },
      ],
      "term": "cat",
    },
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "un": {
              "definitions": [
                {
                  "value": "a dog",
                },
              ],
            },
          },
        },
      ],
      "pronunciation": "dooooog",
      "term": "dog",
    },
  ],
]
`;

exports[`Dictionary lookup considers markdown strategies correctly 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "v": {
              "definitions": [
                {
                  "value": "This **is** a _markdown_ test",
                },
              ],
            },
          },
        },
      ],
      "term": "markdown",
    },
  ],
]
`;

exports[`Dictionary lookup considers markdown strategies correctly 2`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "v": {
              "definitions": [
                {
                  "value": "This is a markdown test",
                },
              ],
            },
          },
        },
      ],
      "term": "markdown",
    },
  ],
]
`;

exports[`Dictionary lookup considers markdown strategies correctly 3`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "v": {
              "definitions": [
                {
                  "value": "This <strong>is</strong> a <em>markdown</em> test",
                },
              ],
            },
          },
        },
      ],
      "term": "markdown",
    },
  ],
]
`;

exports[`Dictionary lookup looks up terms properly 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    "There goes a cat!",
                  ],
                  "notes": [
                    {
                      "examples": [
                        "Some example",
                      ],
                      "value": "Some definition note",
                    },
                  ],
                  "value": "a cat",
                },
              ],
            },
          },
        },
      ],
      "term": "cat",
    },
  ],
]
`;

exports[`Dictionary restarts if the process was killed 1`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "un": {
              "definitions": [
                {
                  "value": "a dog",
                },
              ],
            },
          },
        },
      ],
      "pronunciation": "dooooog",
      "term": "dog",
    },
  ],
]
`;

exports[`Dictionary restarts if the process was killed 2`] = `
[
  [
    {
      "etymologies": [
        {
          "description": "Latin root",
          "senses": {
            "n": {
              "definitions": [
                {
                  "examples": [
                    "The dog runs after the cat.",
                    "The horse ran away.",
                  ],
                  "value": "(vertebrates) To move swiftly",
                },
                {
                  "value": "Act or instance of hurrying (to or from a place) (not necessarily by foot); dash or errand, trip.",
                },
                {
                  "value": "A pleasure trip.",
                },
                {
                  "value": "Flight, instance or period of fleeing.",
                },
                {
                  "value": "Migration (of fish).",
                },
                {
                  "value": "A group of fish that migrate, or ascend a river for the purpose of spawning.",
                },
              ],
            },
            "v": {
              "groups": [
                {
                  "definitions": [
                    {
                      "examples": [
                        "The dog runs after the cat.",
                        "The horse ran away.",
                      ],
                      "value": "(vertebrates) To move swiftly (figuratively)",
                    },
                    {
                      "value": "(fluids) To flow.",
                    },
                    {
                      "value": "(nautical, of a vessel) To sail before the wind, in distinction from reaching or sailing close-hauled.",
                    },
                    {
                      "value": "(social) To carry out an activity.",
                    },
                    {
                      "value": "To extend or persist, statically or dynamically, through space or time.",
                    },
                    {
                      "value": "(transitive) To execute or carry out a plan, procedure or program.",
                    },
                  ],
                  "description": "A number of verb senses",
                },
              ],
            },
          },
        },
      ],
      "term": "run",
    },
  ],
]
`;
